src/rdkafka.h

```diff
@@ -1214,11 +1214,29 @@ RD_EXPORT size_t rd_kafka_header_cnt (const rd_kafka_headers_t *hdrs);
  * @brief Configuration result type
  */
 typedef enum {
-	RD_KAFKA_CONF_UNKNOWN = -2, /**< Unknown configuration name. */
-	RD_KAFKA_CONF_INVALID = -1, /**< Invalid configuration value. */
-	RD_KAFKA_CONF_OK = 0        /**< Configuration okay */
+	RD_KAFKA_CONF_SCOPE_MISMATCH = -3, /**< Invalid scope for consumer or
+                                             * producer configuration option */
+	RD_KAFKA_CONF_UNKNOWN = -2,        /**< Unknown configuration name. */
+	RD_KAFKA_CONF_INVALID = -1,        /**< Invalid configuration value. */
+	RD_KAFKA_CONF_OK = 0               /**< Configuration okay */
 } rd_kafka_conf_res_t;
 
+/**
+ * @brief Enables or disables scope validation on all handle and topic
+ *        configuration settings
+ * @param enable 0 to disable and 1 to enable. Default is disabled.
+ * @remark This feature validates configuration settings and only allows
+ *         appropriate settings within the same scope to be applied
+ *         together. For example one cannot mix consumer and producer settings
+ *         in the same configuration or topic configuration handle. Similar
+ *         checks are performed when setting callback methods which prevents
+ *         users from setting a consumer callback on a producer (and vice-versa)
+ *         as it would have no effect.
+ * @remark The type of scope of a configuration is determined by the first
+ *         non-global option which is set. Any subsequent options must be within
+ *         the same scope.
+ */
+void rd_kafka_conf_enable_scope_validation(int enable);
```

Comment:
This needs to be set per rd_kafka_conf_t and rd_kafka_topic_conf_t, not globally.

Reply:
Yes but why do it for each? It's a feature you want for all configs or none at all. Users will have to call this for every config object they create including the default ones, it's very redundant. This is supposed to be a universal check in place. And then if you miss one, you risk again setting the wrong config options.

Reply:
As a user of librdkafka, how can you know that your use of rdkafka is the only use in the current application runtime? Unless you know exactly what all of your dependencies, and their dependencies, and their dependencies do, you simply can't. This means that each rdkafka client instance must be completely separated from any other instance in the same process.

Reply:
Typically applications would be multi threaded and they would link once the .a or .so of librdkafka. Settings will be done once and you control how APIs are called. I can't think of any use case where you would be, inside your own librdkafka application, linking some other 3rd party application which also happens to be a librdkafka client? or link your .so file? This is not a library like boost or STL where calls need to be completely re-entrant. The config settings array you have can be manipulated in the same way, so I don't see how it's any different. The extern declaration has only internal exposure, nothing leaks outside.

Reply:
I think @edenhill refers to the case where you use a third party library that also uses librdkafka, then you enable this globally and that library has e.g. a mix of producer/consumer configs in the same handle. Then that library will think something's wrong when setting those attributes and possibly fail to work because of that, when in reality it's just that you asked librdkafka to be more strict than it is by default and that as a side effect broke the third party library.

Reply:
ok i'll remove the global setting